### ISSUE TITLE
[5.4.13] Harmony Patcher is throwing Exception

### ISSUE THREAD

**Issue by BlazingTwist** [BepInEx/BepInEx#286]
### What happened?

When patching a certain method in Streets Of Rogue (`LoadLevel::CreateInitialMap()`), Harmony will throw an exception.
Notably, this does not occur when using BepInEx 5.4.11

My assumption is that there is a sanity check being executed that the original method does not satisfy. (The game's code is a bit of a mess.)

### Steps to reproduce

Start game with BepInEx 5.4.13, no patchers, only plugin is the Example-Mod.
[HarmonyException.txt](https://github.com/BepInEx/BepInEx/files/6904352/HarmonyException.txt)

The relevant (decompiled) method has been attached as `LoadLevel.txt`
[LoadLevel.txt](https://github.com/BepInEx/BepInEx/files/6904353/LoadLevel.txt)

### BepInEx Distribution

Stable from GitHub

### Log outputs

Observed log:
```txt
[Log]	[2:04:01 AM]
Example Mod calls Harmony.PatchAll

[Exception]	[2:04:01 AM]
ArgumentNullException: Value cannot be null.
Parameter name: declaringType
Provided Trace: Mono.Cecil.Mixin.CheckType (System.Object type, Mono.Cecil.Mixin+Argument argument) (at <6034b380a22b41a596c9dc29d282c0a9>:0)
Mono.Cecil.MethodReference..ctor (System.String name, Mono.Cecil.TypeReference returnType, Mono.Cecil.TypeReference declaringType) (at <6034b380a22b41a596c9dc29d282c0a9>:0)
MonoMod.Utils.Extensions.Relink (Mono.Cecil.MethodReference method, MonoMod.Utils.Relinker relinker, Mono.Cecil.IGenericParameterProvider context) (at <041d70ff506f4f089a67adab0245e45d>:0)
MonoMod.Utils.Extensions.Relink (Mono.Cecil.IMetadataTokenProvider mtp, MonoMod.Utils.Relinker relinker, Mono.Cecil.IGenericParameterProvider context) (at <041d70ff506f4f089a67adab0245e45d>:0)
MonoMod.Utils.DMDCecilGenerator._Generate (MonoMod.Utils.DynamicMethodDefinition dmd, System.Object context) (at <041d70ff506f4f089a67adab0245e45d>:0)
MonoMod.Utils.DMDGenerator`1[TSelf].Generate (MonoMod.Utils.DynamicMethodDefinition dmd, System.Object context) (at <041d70ff506f4f089a67adab0245e45d>:0)
MonoMod.Utils.DynamicMethodDefinition.Generate (System.Object context) (at <041d70ff506f4f089a67adab0245e45d>:0)
MonoMod.Utils.DynamicMethodDefinition.Generate () (at <041d70ff506f4f089a67adab0245e45d>:0)
DMD<Refresh>?799369984._MonoMod_RuntimeDetour_ILHook+Context::Refresh (MonoMod.RuntimeDetour.ILHook+Context this) (at <70e7062c02c4435ea2f41c18db8e71f9>:0)
DMD<>?799369984.Trampoline<MonoMod.RuntimeDetour.ILHook+Context::Refresh>?-2003272704 (System.Object ) (at <1559e43363754f32815649b54ef64c2d>:0)
HarmonyLib.Internal.RuntimeFixes.StackTraceFixes.OnILChainRefresh (System.Object self) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
MonoMod.RuntimeDetour.ILHook.Apply () (at <45d5519965ed40d2af73eecdf7fbc188>:0)
HarmonyLib.Public.Patching.ManagedMethodPatcher.DetourTo (System.Reflection.MethodBase replacement) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
Rethrow as HarmonyException: IL Compile Error (unknown location)
HarmonyLib.Public.Patching.ManagedMethodPatcher.DetourTo (System.Reflection.MethodBase replacement) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyLib.PatchFunctions.UpdateWrapper (System.Reflection.MethodBase original, HarmonyLib.PatchInfo patchInfo) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
Rethrow as HarmonyException: IL Compile Error (unknown location)
HarmonyLib.PatchClassProcessor.ReportException (System.Exception exception, System.Reflection.MethodBase original) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyLib.PatchClassProcessor.Patch () (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyLib.Harmony.<PatchAll>b__11_0 (System.Type type) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyLib.CollectionExtensions.Do[T] (System.Collections.Generic.IEnumerable`1[T] sequence, System.Action`1[T] action) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyLib.Harmony.PatchAll (System.Reflection.Assembly assembly) (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyLib.Harmony.PatchAll () (at <225ac1e91b3845d9bbd0b8e0ff38fa08>:0)
HarmonyPatchallException.HarmonyPatchAllException.Awake () (at <c4cc17343388413d9bb7a9ddde9903d5>:0)
UnityEngine.GameObject:AddComponent(Type)
BepInEx.Bootstrap.Chainloader:Start()
UnityEngine.Application:.cctor()
Rewired.InputManager_Base:Awake()
```

Expected Log
```txt
[Log]	[2:04:01 AM]
Example Mod calls Harmony.PatchAll

[Log]	[2:04:01 AM]
Example Mod finished without exception.
```

### Environment

```markdown
- OS: Windows 10 x64
- BepInEx: 5.4.13
- Game: Streets of Rogue v95 x64
```

**Comment by ghorsington** [comment]
Greetings!

Thank you for the extensive info you provided! Thanks to it I managed to reproduce your error.

It looks like it's an issue on MonoMod's side. From the looks of it, the issue was fixed recently in MonoMod but the update hasn't been included in BepInEx 5.4.13. I'll update BepInEx to include the fix and then close this issue within the next 24 hours. 

Meanwhile, you can use the following temporary workaround if you absolutely need to:

1. Download latest MonoMod: [MonoMod-21.07.22.03-net35.zip](https://github.com/MonoMod/MonoMod/releases/download/v21.07.22.03/MonoMod-21.07.22.03-net35.zip)
2. From the ZIP, take `MonoMod.RuntimeDetour.dll` and `MonoMod.Utils.dll` and place them into `BepInEx/core` folder. Overwrite the files when asked.
3. Run the game normally. The game will run without errors and the prefix will be called.

As an example, here is my test plugin which is based on your example: 
[TestException.zip](https://github.com/BepInEx/BepInEx/files/6905137/TestException.zip)

With the workaround it runs correctly and outputs a test message on level load:
![image](https://user-images.githubusercontent.com/13762037/127602667-2959f8d2-cddd-4eae-b4a8-b66493234f20.png)

**Comment by BlazingTwist** [comment]
Very nice. Thanks for the quick reply.

**Comment by ghorsington** [comment]
This is now fixed in BepInEx 5.4.14. As such, I'm closing the issue.
